pyfstat/snr.py

```diff
@@ -0,0 +1,423 @@
+
+        sky = lal.SkyPosition()
+        sky.longitude = Alpha
+        sky.latitude = Delta
+        sky.system = lal.COORDINATESYSTEM_EQUATORIAL
+        lal.NormalizeSkyPosition(sky.longitude, sky.latitude)
+
+        Mmunu = lalpulsar.ComputeMultiAMCoeffs(
+            multiDetStates=self.detector_states,
+            multiWeights=self.noise_weights,
+            skypos=sky,
+        ).Mmunu
+
+        if self.noise_weights is None:
+            Mmunu.Sinv_Tsft = self.Sinv_Tsft
+
+        return Mmunu
+
+    def _convert_amplitude_parameters(self, h0, cosi, aPlus, aCross):
+        """
+        Internal method to check and convert the given amplitude parameters
+        into the required format.
+        """
+        h0_cosi = h0 is not None and cosi is not None
+        aPlusCross = aPlus is not None and aCross is not None
+
+        if h0_cosi == aPlusCross:
+            raise ValueError("Need either (h0, cosi) or (aPlus, aCross), but not both")
+
+        if h0_cosi:
+            aPlus = 0.5 * h0 * (1 + cosi**2)
+            aCross = h0 * cosi
+            return aPlus, aCross
+
+        return aPlus, aCross
+
+
+class DetectorStates:
+    """
+    Python interface to XLALGetMultiDetectorStates and XLALGetMultiDetectorStatesFromMultiSFTs.
+    """
+
+    def __init__(self):
+        self.ephems = lalpulsar.InitBarycenter(*get_ephemeris_files())
+
+    def multi_detector_states(self, timestamps, Tsft, detectors=None, time_offset=0):
+        """
+        Parameters
+        ----------
+        time_stamps: array-like or dict
+            GPS timestamps at which detector states will be retrieved.
+            If array, use the same set of timestamps for all detectors,
+            which must be explicitly given by the user via `detectors`.
+            If dictionary, each key should correspond to a valid detector name
+            to be parsed by XLALParseMultiLALDetector and the associated value
+            should be an array-like set of GPS timestamps for each individual detector.
+        Tsft: float
+            Timespan covered by each timestamp. I does not need to coincide with the
+            separation between consecutive timestamps.
+        detectors: list[str] or comma-separated string
+            List of detectors to be parsed using XLALParseMultiLALDetector.
+            Conflicts with dictionary of `time_stamps`, required otherwise.
+        time_offset: float
+            Timestamp offset to retrieve detector states.
+
+
+        Returns
+        -------
+        multi_detector_states: lalpulsar.MultiDetectorStateSeries
+            Resulting multi-detector states produced by XLALGetMultiDetectorStates
+        """
+
+        self._parse_timestamps_and_detectors(timestamps, Tsft, detectors)
+        return lalpulsar.GetMultiDetectorStates(
+            self.multi_timestamps,
+            self.multi_detector,
+            self.ephems,
+            time_offset,
+        )
+
+    def multi_detector_states_from_sfts(
```

Comment:
Didn't catch this on first round, but this should probably better be called `get_multi_detector_states_from_sfts`.